src/integrity-checker.js

```diff
@@ -118,23 +109,59 @@ export default class InstallationIntegrityChecker {
   }
 
   /**
-   * returns a list of files recursively in a directory sorted
+   * Get the list of the directories that contain our modules (there might be multiple such folders b/c of workspaces).
    */
-  async _getFilesDeep(rootDir: string): Promise<Array<string>> {
-    async function getFilePaths(rootDir: string, files: Array<string>, currentDir: string = rootDir): Promise<void> {
-      for (const file of await fs.readdir(currentDir)) {
-        const entry = path.join(currentDir, file);
+
+  _getModulesFolders({workspaceLayout}: {workspaceLayout: ?WorkspaceLayout} = {}): Array<string> {
+    const locations = [];
+
+    if (this.config.modulesFolder) {
+      locations.push(this.config.modulesFolder);
+    } else {
+      locations.push(path.join(this.config.lockfileFolder, 'node_modules'));
+    }
+
+    if (workspaceLayout) {
+      for (const workspaceName of Object.keys(workspaceLayout.workspaces)) {
+        const loc = workspaceLayout.workspaces[workspaceName].loc;
+
+        if (loc) {
+          locations.push(path.join(loc, 'node_modules'));
+        }
+      }
+    }
+
+    return locations;
+  }
+
+  /**
+   * Get a list of the files that are located inside our module folders.
+   */
+  async _getIntegrityListing({workspaceLayout}: {workspaceLayout: ?WorkspaceLayout} = {}): Promise<Array<string>> {
+    const files = [];
+
+    const recurse = async dir => {
+      if (!await fs.exists(dir)) {
+        return;
+      }
+
+      for (const file of await fs.readdir(dir)) {
+        const entry = path.join(dir, file);
         const stat = await fs.lstat(entry);
+
         if (stat.isDirectory()) {
-          await getFilePaths(rootDir, files, entry);
+          await recurse(entry);
         } else {
-          files.push(path.relative(rootDir, entry));
+          files.push(entry);
         }
       }
+    };
+
+    for (const modulesFolder of this._getModulesFolders({workspaceLayout})) {
+      await recurse(modulesFolder);
```

Comment:
No need to serialize each of these. Why not use `Promise.all` here:
```
const allFolders = [];
for (const modulesFolder of this._getModulesFolders({workspaceLayout})) {
  allFolders.push(recurse(modulesFolder));
});
await Promise.all(allFolders);
```

Alternatively:
```
await Promise.all(this._getModulesFolders({workspaceLayout}).map(folder => recuse(folder)));
```

Also, instead of changing `files` implicitly, it may be better to return it from `recurse` and then `.concat()` at the end. May be slower and churn memory tho.